Update required libyuv version

ArchangeGabriel opened the pull request:
Fixes GH-781 (the function was introduced in libavif https://github.com/AOMediaCodec/libavif/commit/46104d6db42ec2a94a7da6fc0693f24fa90a1b98).

wantehchang commented on CMakeLists.txt:
```diff
@@ -225,10 +225,10 @@ endif()
 find_package(libyuv QUIET) # not required
 if(libyuv_FOUND)
-    # libyuv 1755 exposed all of the I*Matrix() functions, which libavif relies on.
+    # libyuv 1778 exposed ScalePlane_12 function, which libavif relies on.
```
libyuv 1778 exposed I420Scale_12 and I444Scale_12, not ScalePlane12:
https://chromium-review.googlesource.com/c/libyuv/libyuv/+/2706966

ScalePlane_12 was added in libyuv 1774:
https://chromium-review.googlesource.com/c/libyuv/libyuv/+/2658097

It is certainly fine to require a newer version such as libyuv 1778. I am just wondering if you know something wrong with the ScalePlane_12() function in libyuv 1774, such as it was an incomplete or buggy implementation.

tongyuantongyu replied:
I recommended to use 1789:
There's some issues in the neon64 assembly when building with gcc:
https://chromium-review.googlesource.com/c/libyuv/libyuv/+/2922147

ArchangeGabriel replied:
@wantehchang You are right, actually I only did quick commit msg search, and Scale 12 matched on https://chromium.googlesource.com/libyuv/libyuv/+/08815a29766a78398a8e2b9ed095280e9d0a73c2^!/. Grepping the commit diff for `ScalePlane_12` was further positive, and this commit allowed libavif to compile, so I thought this was good. But indeed looking more deeply at the diff I now see that `ScalePlane_12` is called but not defined in this commit (which is as you said about `I420/I444Scale_12`), you are evidently right that 1774 should be enough. But I can update the MR to 1789 instead to avoid the issue mentionned by @tongyuantongyu.

ArchangeGabriel replied:
Apparently they are other neon fixes in 1796: https://chromium-review.googlesource.com/c/libyuv/libyuv/+/3216739

wantehchang reviewed (commented):
Hi Bruno,

Thank you very much for the bug report and the pull request. I have a question.

wantehchang commented on CMakeLists.txt:
```diff
@@ -225,10 +225,10 @@ endif()
     if(NOT LIBYUV_VERSION)
         message(STATUS "libavif: libyuv found, but version unknown; libyuv-based fast paths disabled.")
-    elseif(LIBYUV_VERSION LESS 1755)
+    elseif(LIBYUV_VERSION LESS 1778)
```
Hi Bruno,

I just realized that this approach has a serious downside: it will cause libavif to not use libyuv for RGB-to-YUV and YUV-to-RGB conversions if libyuv is not version 1778 or newer.

libavif uses libyuv for two purposes: RGB<->YUV conversion and scaling. Between the two, RGB<->YUV conversion is much more important.

I suggest we use a different approach. We check for libyuv version at the place where we call ScalePlane_12(). Please search for "LIBYUV_VERSION" in libavif/src/reformat_libyuv.c for examples.

I think something like this patch should work. In this patch I use libyuv version 1774 for simplicity and leave the responsibility of using a libyuv version that is good enough to the user.

```
diff --git a/src/scale.c b/src/scale.c
index 6ca651b..89bf2fe 100644
--- a/src/scale.c
+++ b/src/scale.c
@@ -71,6 +71,13 @@ avifBool avifImageScale(avifImage * image, uint32_t dstWidth, uint32_t dstHeight
     image->height = dstHeight;
 
     if (srcYUVPlanes[0] || srcAlphaPlane) {
+#if LIBYUV_VERSION < 1774
+        if (image->depth > 8) {
+            avifDiagnosticsPrintf(diag, "ScalePlane_12() is not available in libyuv version %d", LIBYUV_VERSION);
+            return AVIF_FALSE;
+        }
+#endif
+
         // A simple conservative check to avoid integer overflows in libyuv's ScalePlane() and
         // ScalePlane_12() functions.
         if (srcWidth > 16384) {
@@ -103,11 +110,13 @@ avifBool avifImageScale(avifImage * image, uint32_t dstWidth, uint32_t dstHeight
             const uint32_t dstW = (i == AVIF_CHAN_Y) ? dstWidth : dstUVWidth;
             const uint32_t dstH = (i == AVIF_CHAN_Y) ? dstHeight : dstUVHeight;
             if (image->depth > 8) {
+#if LIBYUV_VERSION >= 1774
                 uint16_t * const srcPlane = (uint16_t *)srcYUVPlanes[i];
                 const uint32_t srcStride = srcYUVRowBytes[i] / 2;
                 uint16_t * const dstPlane = (uint16_t *)image->yuvPlanes[i];
                 const uint32_t dstStride = image->yuvRowBytes[i] / 2;
                 ScalePlane_12(srcPlane, srcStride, srcW, srcH, dstPlane, dstStride, dstW, dstH, AVIF_LIBYUV_FILTER_MODE);
+#endif
             } else {
                 uint8_t * const srcPlane = srcYUVPlanes[i];
                 const uint32_t srcStride = srcYUVRowBytes[i];
@@ -126,11 +135,13 @@ avifBool avifImageScale(avifImage * image, uint32_t dstWidth, uint32_t dstHeight
         avifImageAllocatePlanes(image, AVIF_PLANES_A);
 
         if (image->depth > 8) {
+#if LIBYUV_VERSION >= 1774
             uint16_t * const srcPlane = (uint16_t *)srcAlphaPlane;
             const uint32_t srcStride = srcAlphaRowBytes / 2;
             uint16_t * const dstPlane = (uint16_t *)image->alphaPlane;
             const uint32_t dstStride = image->alphaRowBytes / 2;
             ScalePlane_12(srcPlane, srcStride, srcWidth, srcHeight, dstPlane, dstStride, dstWidth, dstHeight, AVIF_LIBYUV_FILTER_MODE);
+#endif
         } else {
             uint8_t * const srcPlane = srcAlphaPlane;
             const uint32_t srcStride = srcAlphaRowBytes;
```

tongyuantongyu replied:
Alternatively, we can use `ScalePlane_16` instead if `LIBYUV_VERSION < 1774`.
`ScalePlane_12` has special optimization for 2x scale up using linear/bilinear filter (when scaling up, box and bilinear are the same), and for the rest cases, it simply calls `ScalePlane_16`, which handles 2x scale up just fine (slower though).

ArchangeGabriel replied:
Seems a good idea. However I’m not exactly sure whether the arguments for `ScalePlane_16` are the same as for the `_12` versions, and especially I don’t know (the) code enough to avoid issues such as https://github.com/AOMediaCodec/libavif/issues/715.

tongyuantongyu replied:
The arguments are the same: https://chromium.googlesource.com/libyuv/libyuv/+/a04e4f87fbf40405707b1d0ae9fcba8fc93f7856/source/scale.cc#2073, so it should be fine to use `ScalePlane_16` when `ScalePlane_12` is unavailable.

ArchangeGabriel commented:
I’ve changed the commit to conditionally use ScalePlane_12 as suggested, and modified CMake to print additional information instead when in the partially supported version range. I can change the message (e.g. specify what can of optimizations, wording. etc.), remove it (but I think people compiling should be told about it so they can leverage it) or merge it with the enabling line (so making to enable message depending on version).

wantehchang reviewed (approved):
LGTM. Thanks.

Bruno: I am very sorry for the late review. I was out of the office.

ArchangeGabriel commented:
It’s fine, no need to apologize. ;)
